connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/IsolatedPlugin.java

```diff
@@ -0,0 +1,122 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.runtime.isolation;
+
+import java.util.Objects;
+import java.util.concurrent.Callable;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+public abstract class IsolatedPlugin<P> {
+
+    private final Plugins plugins;
+    private final Class<?> pluginClass;
+    protected final P delegate;
+    private final ClassLoader classLoader;
+    private final PluginType type;
+
+    IsolatedPlugin(Plugins plugins, P delegate, PluginType type) {
+        this.plugins = Objects.requireNonNull(plugins, "plugins must be non-null");
+        this.delegate = Objects.requireNonNull(delegate, "delegate plugin must be non-null");
+        this.pluginClass = delegate.getClass();
+        ClassLoader classLoader = pluginClass.getClassLoader();
+        this.classLoader = Objects.requireNonNull(classLoader, "delegate plugin must not be a boostrap class");
+        this.type = Objects.requireNonNull(type, "plugin type must be non-null");
+    }
+
+    public PluginType type() {
+        return type;
+    }
+
+    @SuppressWarnings("unchecked")
+    public Class<? extends P> pluginClass() {
+        return (Class<? extends P>) pluginClass;
+    }
+
+    protected <V> V isolate(Callable<V> callable) throws Exception {
+        try (LoaderSwap loaderSwap = plugins.withClassLoader(classLoader)) {
+            return callable.call();
+        }
+    }
+
+    protected void isolateV(ThrowingRunnable runnable) throws Exception {
+        try (LoaderSwap loaderSwap = plugins.withClassLoader(classLoader)) {
+            runnable.run();
+        }
+    }
+
+    protected <T> void isolateV(Consumer<T> consumer, T t) throws Exception {
+        try (LoaderSwap loaderSwap = plugins.withClassLoader(classLoader)) {
+            consumer.accept(t);
+        }
+    }
+
+    protected <T, U> void isolateV(BiConsumer<T, U> consumer, T t, U u) throws Exception {
+        try (LoaderSwap loaderSwap = plugins.withClassLoader(classLoader)) {
+            consumer.accept(t, u);
+        }
+    }
+
+    protected <T, R> R isolate(Function<T, R> function, T t) throws Exception {
+        try (LoaderSwap loaderSwap = plugins.withClassLoader(classLoader)) {
+            return function.apply(t);
+        }
+    }
+
+    protected <T, U, R> R isolate(BiFunction<T, U, R> function, T t, U u) throws Exception {
+        try (LoaderSwap loaderSwap = plugins.withClassLoader(classLoader)) {
+            return function.apply(t, u);
+        }
+    }
```

Comment:
I think this is a question of readability, but it's important to consider the readability for operators of these connectors. With the current design that uses mostly method references, exceptions have the following stacktrace:
```
org.apache.kafka.connect.errors.ConnectException:
	at org.apache.kafka.connect.runtime.isolation.IsolatedPluginTest.lambda$testSinkConnector$0(IsolatedPluginTest.java:43)
        ...
	at org.apache.kafka.connect.connector.Connector.initialize(Connector.java:57)
	at org.apache.kafka.connect.runtime.isolation.IsolatedPlugin.isolateV(IsolatedPlugin.java:66)
	at org.apache.kafka.connect.runtime.isolation.IsolatedConnector.initialize(IsolatedConnector.java:39)
	at org.apache.kafka.connect.runtime.isolation.IsolatedPluginTest.testSinkConnector(IsolatedPluginTest.java:46)
	...
```

There's two stack frames associated with the isolation infrastructure that did not exist befor, and neither of them are anonymous lambdas (the `IsolatedPluginTest.lambda$testSinkConnector$0` is an anonymous lambda in my test, not part of the IsolatedPlugin).

Compare that to the style of making everything into a callable:
```
org.apache.kafka.connect.errors.ConnectException:
	at org.apache.kafka.connect.runtime.isolation.IsolatedPluginTest.lambda$testSinkConnector$0(IsolatedPluginTest.java:43)
        ...
	at org.apache.kafka.connect.connector.Connector.initialize(Connector.java:57)
	at org.apache.kafka.connect.runtime.isolation.IsolatedConnector.lambda$initialize$0(IsolatedConnector.java:39)
	at org.apache.kafka.connect.runtime.isolation.IsolatedPlugin.isolateV(IsolatedPlugin.java:60)
	at org.apache.kafka.connect.runtime.isolation.IsolatedConnector.initialize(IsolatedConnector.java:39)
	at org.apache.kafka.connect.runtime.isolation.IsolatedPluginTest.testSinkConnector(IsolatedPluginTest.java:46)
	...
```

This has one additional stacktrace for the lambda mapping from the function to a callable, and it has a bit of an ugly name.

Also it doesn't appear in this PR, but I did find some plugin methods which cannot be reduced to one of these signatures, and did have to be reduced to a Callable (I think it was the converters?).
So this strategy is not perfect, does not cover all of the cases, but does provide slightly cleaner stacktraces in the situations it does cover.
I think i could go either way on this.

Reply:
To take it to the extreme, we would inline and eliminate each of the `isolate` calls entirely and put the try-with-resources swap in the methods themselves. That would eliminate all lambdas and only add one stack frame, but be super repetitive in the high 10s of methods that have to be isolated.

Reply:
I would be really interested in knowing which case can't be covered here. If you can dig up the `Converter` counterexample, would you mind sharing it?

Reply:
If it comes down to not being able to completely satisfy all use cases with just the `isolate(Callable)` (and possibly `isolateV(ThrowingRunnable)` methods, then I think it could be alright to add more variants. But the additional stack frames, and even ugly class names, don't seem likely to be super important to users or plugin developers. In both cases (though probably more often the latter), aren't they more likely to look for the first stack frame that originates with their plugin class?

Reply:
> I would be really interested in knowing which case can't be covered here. If you can dig up the Converter counterexample, would you mind sharing it?

Here's a snippet from the IsolatedConverter, which is a few PRs in the future from now.
These methods take 3 or 4 arguments, and there's no builtin java functional type that covers them (I guess they would be TriFunction and QuadFunction?)
```
    public byte[] fromConnectData(String topic, Headers headers, Schema schema, Object value) throws Exception {
        return isolate(() -> delegate.fromConnectData(topic, headers, schema, value));
    }

    public SchemaAndValue toConnectData(String topic, Headers headers, byte[] value) throws Exception {
        return isolate(() -> delegate.toConnectData(topic, headers, value));
    }

    public byte[] fromConnectData(String topic, Schema schema, Object value) throws Exception {
        return isolate(() -> delegate.fromConnectData(topic, schema, value));
    }
    
    public SchemaAndValue toConnectData(String topic, byte[] value) throws Exception {
        return isolate(delegate::toConnectData, topic, value);
    }
```

The

Reply:
Ah, I misunderstood--I thought you'd found a counterexample that couldn't be covered by the minimal set of `isolate`/`isolateV` variants, instead of one that couldn't be covered by the expanded set of variants originally present in the PR. Was really scratching my head over that one!

Thanks for the changes, this class LGTM now 👍